Library/Homebrew/extend/fileutils.rb

```diff
@@ -6,42 +6,79 @@
 # @see http://ruby-doc.org/stdlib-1.8.7/libdoc/fileutils/rdoc/FileUtils.html Ruby's FileUtils API
 module FileUtils
   # Create a temporary directory then yield. When the block returns,
-  # recursively delete the temporary directory.
-  def mktemp(prefix = name)
-    prev = pwd
-    tmp  = Dir.mktmpdir(prefix, HOMEBREW_TEMP)
-
-    # Make sure files inside the temporary directory have the same group as the
-    # brew instance.
-    #
-    # Reference from `man 2 open`
-    # > When a new file is created, it is given the group of the directory which
-    # contains it.
-    group_id = if HOMEBREW_BREW_FILE.grpowned?
-      HOMEBREW_BREW_FILE.stat.gid
-    else
-      Process.gid
+  # recursively delete the temporary directory. Passing opts[:preserve]
+  # or calling `do |stg| ... stg.preserve` in the block will suppress
+  # the deletion for the topmost mktemp on the call stack, if any.
+  def mktemp(prefix = name, opts = {})
+    Mktemp.new(prefix, opts).run do |stg|
+      yield stg
     end
-    begin
-      # group_id.to_s makes OS X 10.6.7 (ruby-1.8.7-p174) and earlier happy.
-      chown(nil, group_id.to_s, tmp)
-    rescue Errno::EPERM
-      opoo "Failed setting group \"#{Etc.getgrgid(group_id).name}\" on #{tmp}"
+  end
+
+  module_function :mktemp
+
+  # Performs mktemp's functionality, and tracks the results.
+  # Each instance is only intended to be used once.
+  class Mktemp
+    include FileUtils
+
+    # Path to the tmpdir used in this run, as a Pathname.
+    attr_accessor :tmpdir
+    # Whether the files in tmpdir should be retained
+    attr_accessor :do_preserve
```

Comment:
`@do_preserve` does not need to be publicly accessible. It is already exposed via `preserve` and `preserve?`.